src/Middleware/OutputCaching/src/Memory/MemoryOutputCacheStore.cs

```diff
@@ -71,26 +93,61 @@ public ValueTask SetAsync(string key, byte[] value, string[]? tags, TimeSpan val
                     keys.Add(key);
                 }
 
-                SetEntry();
+                SetEntry(key, value, tags, validFor);
             }
         }
         else
         {
-            SetEntry();
+            SetEntry(key, value, tags, validFor);
         }
 
-        void SetEntry()
+        return ValueTask.CompletedTask;
+    }
+
+    void SetEntry(string key, byte[] value, string[]? tags, TimeSpan validFor)
+    {
+        var options = new MemoryCacheEntryOptions
         {
-            _cache.Set(
-            key,
-            value,
-            new MemoryCacheEntryOptions
-            {
-                AbsoluteExpirationRelativeToNow = validFor,
-                Size = value.Length
-            });
+            AbsoluteExpirationRelativeToNow = validFor,
+            Size = value.Length
+        };
+
+        if (tags != null && tags.Length > 0)
+        {
+            // Remove cache keys from tag lists when the entry is evicted
+            options.RegisterPostEvictionCallback(RemoveFromTags, tags);
         }
 
-        return ValueTask.CompletedTask;
+        _cache.Set(key, value, options);
+    }
+
+    void RemoveFromTags(object key, object? value, EvictionReason reason, object? state)
+    {
+        var tags = state as string[];
+
+        Debug.Assert(tags != null);
+        Debug.Assert(tags.Length > 0);
+
+        if (key is not string stringKey)
+        {
+            return;
+        }
+
+        lock (_tagsLock)
+        {
+            foreach (var tag in tags!)
+            {
+                if (_taggedEntries.TryGetValue(tag, out var tagged) && tagged != null)
```

Comment:
Would we ever actually put null in the `HashSet<string>`? I think we should throw an `ArgumentException` from `SetAsync` given a null tag in a non-null tags array.